rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,839 @@
+- [OCaml's representation of closures](https://github.com/Gbury/ocaml-memgraph/blob/master/doc/closures.md)
+
+### Haskell
+
+Despite not being advertised, Haskell has an interpreter as well, which is used
+mostly for the GHCi REPL. This section describes what we know of the actual the
+bytecode interpreter, but also incorporates some aspects of the original STG
+machine which is the basis of the low-level operational semantics of Haskell.
+
+#### Memory representation
+
+Closures (thunks and functions) are represented uniformly as an environment and
+a pointer to a closure info table. The closure info table contains the code
+pointer and other garbage-collection related data. The closure info table is
+generated only once per binding appearing in the code and thus shared by dynamic
+instances of the same code. In some other machines for lazy lambda calculi,
+thunks hold additional data such as a flag (evaluated or suspended) and a
+potential value that is filled after the thunk has been forced. When accessing a
+thunk, the caller checks the state, and then either retrieve the value or
+initiate an update sequence and retrieve the code of the unevaluated expression.
+In Haskell, the update process is performed _by the callee_ instead (the thunk's
+code), such that thunk access is uniform: it's an unconditional jump to the
+corresponding code.
+
+As with other VMs, the environment -- immediately inlined after the code pointer
+-- stores the free variables of the expression that are captured by the closure.
+When entering a closure, a dedicated register holds the pointer to the
+environment for fast access.
+
+The STG paper argues that this uniform thunk representation (with "self-handled
+update") simplifies the compilation process and gives room for some specific
+optimizations (vectored return for pattern matching, for example) that should be
+beneficial to Haskell programs.
+
+In Haskell, every data is considered to be an algebraic data type, including
+primitive types such as integers, which is just `data Int = MkInt Int#` (where
+`Int#` is a native machine integer). Indeed, ADTs are ubiquitous in functional
+programming, and should be first-class citizen (regarding performance in
+particular). In consequence, ADTs are the basis of all data values. The
+primitive operation to look at them is the `case` expression (this is a
+low-level case, more primitive than the fancy pattern matching of the surface
+language), which forces its argument.
+
+Both data and closures are laid out in a similar way in memory. A data value is
+a block with a code pointer representing the corresponding ADT constructor. The
+code pointer is followed by the constructor's argument (instead of the
+environment in the case of closures). As each constructor usage potentially
+generates very similar code, GHC is smart enough to share common constructors
+instead of generating them again and again (typically the one for an empty
+list).
```

Comment:
An interesting difference between GHC's memory representation and Ocaml's is the way unboxed and boxed values are distinguished. This is only a concern for the GC, so maybe it doesn't matter too much for this discussion.

- In Ocaml, if a value, viewed as an unsigned int, is odd, then it's unboxed (and the GC doesn't follow it), when it's even, it's a pointer to a boxed value. This way it's always evident, when looking at a value whether it's a pointer or not. The cost is that integers lose one bit, and that it becomes impossible to unbox floats in most cases.
- In Haskell, pointers and non-pointer values are indistinguishable. So the GC asks the info table which of the fields are pointers. In practice, this is just a number, and all the non-pointer fields are stored first, and all the pointer fields after (or maybe the other way around, I don't remember).